crypto/bio/bss_mem.c

```diff
@@ -278,8 +293,9 @@ static long mem_ctrl(BIO *b, int cmd, long num, void *ptr)
         break;
     case BIO_C_GET_BUF_MEM_PTR:
         if (ptr != NULL) {
-            mem_buf_sync(b);
-            bm = bbm->readp;
+            if (!(b->flags & BIO_FLAGS_MEM_RDONLY))
+                mem_buf_sync(b);
+            bm = bbm->buf;
```

Comment:
it is a bit of a conflict, here.
naturally you would expect BIO_get_mem_ptr to return the same BUF_MEM that was set with BIO_set_mem_buf
but you would also expect that the BUF_MEM's data is at the current read position of the BIO.

Reply:
if BIO_NOCLOSE is used, you would expect to be able to use the BUF_MEM from BIO_get_mem_buf
and set it to another BIO with BIO_set_mem_buf, with or without NOCLOSE and/or NONCLEAR_RST.

Reply:
Could it be solved somewhat by documenting it? I mean the RDONLY buf semantics can be hardly made perfect in this sense because there are other conflicting requirements. I tried to make it the other way around first - i.e. to fully switch the meaning of buf and readp for RDONLY BIO but that breaks in case the code is written as in the test_bio_memleak() - i.e. you first call BIO_set_mem_buf() and only after that you set the RDONLY flag.
Of course even this implementation is nothing perfect because if you randomly switch the RDONLY flag on and off between reads and writes all the weird stuff can happen. But that limitation can be documented as well.

Reply:
Yes I too tried to fix it first but lost my way in the woods...
My thinking is that the API was designed in a way how it cannot work,
for instance NONCLEAR_RST I cannot think of a test case where it would ever work.
That was this way right from the first time this was implemented, and till today.
Maybe there is a way how it would do something sensible, but I was unable to find it.
Really, no joke.
Therefore, given that nobody complained so far, I expect that this one has never
been used anywhere, so we could as well drop it.
I would also think that it is acceptable to make BIO_get_mem_buf fail when
RDONLY flag is used.

Reply:
I think the case where you do BIO_get_mem_buf() on RDONLY BIO with BIO_NOCLOSE and set it to another RDONLY BIO should work with this patch, I'll try to add a testcase. I do not dispute that the NONCLEAR_RST is broken and unfixable. I did not even try to think about it more than it would be extremely hard to fix in any meaningful way and it is probably not worth attempting it as it never worked.

Reply:
According to https://codesearch.debian.net/search?q=NONCLEAR_RST, there are no users in Debian of that.